File: src/ReadStorm.Application/Abstractions/IDownloadBookUseCase.cs
using ReadStorm.Domain.Models;

namespace ReadStorm.Application.Abstractions;

public interface IDownloadBookUseCase
{
    Task QueueAsync(
        DownloadTask task,
        SearchResult selectedBook,
        DownloadMode mode,
        CancellationToken cancellationToken = default);

    /// <summary>
    /// 检查书籍是否有新章节。重新抓取目录并与 DB 对比，
    /// 新章节以 Pending 状态插入。返回新增章节数。
    /// </summary>
    Task<int> CheckNewChaptersAsync(
        BookEntity book,
        CancellationToken cancellationToken = default);

    /// <summary>
    /// 换源：从指定书源重新抓取当前章节内容。
    /// 返回 (成功?, 新正文, 诊断信息)。
    /// </summary>
    Task<(bool Success, string Content, string Message)> FetchChapterFromSourceAsync(
        BookEntity book,
        string chapterTitle,
        int sourceId,
        CancellationToken cancellationToken = default);
}

File: pureopus/app/src/main/java/com/readstorm/app/application/abstractions/IDownloadBookUseCase.kt
package com.readstorm.app.application.abstractions

import com.readstorm.app.domain.models.BookEntity
import com.readstorm.app.domain.models.DownloadMode
import com.readstorm.app.domain.models.DownloadTask
import com.readstorm.app.domain.models.SearchResult

interface IDownloadBookUseCase {
    suspend fun queue(task: DownloadTask, selectedBook: SearchResult, mode: DownloadMode)

    suspend fun checkNewChapters(book: BookEntity): Int

    /** Returns (success, content, message). */
    suspend fun fetchChapterFromSource(
        book: BookEntity,
        chapterTitle: String,
        sourceId: Int
    ): Triple<Boolean, String, String>
}
